File: tf2_trajectory_planning_msgs/test/impl/test_tf2_trajectory_planning_msgs.cpp
/** ============================================================================
MIT License

Copyright (c) 2025 Institute for Automotive Engineering (ika), RWTH Aachen University

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
============================================================================= */

#include <cmath>

#include <gtest/gtest.h>

using namespace trajectory_planning_msgs::trajectory_access;

static const double EPS = 1e-12;

TEST(tf2_trajectory_planning_msgs, test_doTransform_REFERENCE) {

  Trajectory trajectory, trajectory_tf;
  initializeTrajectory(trajectory, REFERENCE::TYPE_ID, 2);

  for (int i = 0; i < getSamplePointSize(trajectory); i++) {
    setT(trajectory, i*1.0, i);
    setX(trajectory, i*1.0, i);
    setY(trajectory, 1.0, i);
    setV(trajectory, 1.0, i);
  }

  gm::TransformStamped tf;
  tf.transform.translation.x = 10.0;
  tf.transform.translation.y = 20.0;
  tf.transform.translation.z = 0.0;
  tf.transform.rotation.x = 0.0;
  tf.transform.rotation.y = 0.0;
  tf.transform.rotation.z = sin(M_PI / 4);
  tf.transform.rotation.w = cos(M_PI / 4);

  tf2::doTransform(trajectory, trajectory_tf, tf);

  for (int i = 0; i < getSamplePointSize(trajectory); i++) {

    // transformed state
    EXPECT_NEAR(getX(trajectory_tf, i), 9, EPS);
    EXPECT_NEAR(getY(trajectory_tf, i), i + 20, EPS);
    EXPECT_NEAR(getTheta(trajectory_tf, i), M_PI / 2, EPS);

    // transform-invariant state
    EXPECT_NEAR(getV(trajectory_tf, i), 1.0, EPS);
  }
}


TEST(tf2_trajectory_planning_msgs, test_doTransform_DRIVABLE) {

  Trajectory trajectory, trajectory_tf;
  initializeTrajectory(trajectory, DRIVABLE::TYPE_ID, 2);

  for (int i = 0; i < getSamplePointSize(trajectory); i++) {
    setT(trajectory, 1.0, i);
    setX(trajectory, 1.0, i);
    setY(trajectory, 2.0, i);
    setV(trajectory, 1.0, i);
    setTheta(trajectory, M_PI, i);
    setA(trajectory, 1.0, i);
    setDeltaAck(trajectory, 1.0, i);
    setS(trajectory, 1.0, i);
  }

  gm::TransformStamped tf;
  tf.transform.translation.x = 10.0;
  tf.transform.translation.y = 20.0;
  tf.transform.translation.z = 30.0;
  tf.transform.rotation.x = 0.0;
  tf.transform.rotation.y = 0.0;
  tf.transform.rotation.z = 1.0;
  tf.transform.rotation.w = 0.0;

  tf2::doTransform(trajectory, trajectory_tf, tf);

  for (int i = 0; i < getSamplePointSize(trajectory); i++) {

    // transformed state
    EXPECT_NEAR(getX(trajectory_tf, i), 9.0, EPS);
    EXPECT_NEAR(getY(trajectory_tf, i), 18.0, EPS);
    EXPECT_NEAR(getTheta(trajectory_tf, i), 0.0, EPS);

    // transform-invariant state
    EXPECT_NEAR(getV(trajectory_tf, i), 1.0, EPS);
    EXPECT_NEAR(getA(trajectory_tf, i), 1.0, EPS);
    EXPECT_NEAR(getDeltaAck(trajectory_tf, i), 1.0, EPS);
    EXPECT_NEAR(getS(trajectory_tf, i), 1.0, EPS);
  }
}

TEST(tf2_trajectory_planning_msgs, test_doTransform_DRIVABLERWS) {

  Trajectory trajectory, trajectory_tf;
  initializeTrajectory(trajectory, DRIVABLERWS::TYPE_ID, 2);

  double delta_front = 1.2;
  double delta_rear = 0.2;
  double distance_front_axle = 1.5;
  double distance_rear_axle = 1.9;
  double beta = atan((distance_rear_axle / (distance_front_axle + distance_rear_axle)) * tan(delta_front) 
                  + (distance_front_axle / (distance_front_axle + distance_rear_axle)) * tan(delta_rear));

  for (int i = 0; i < getSamplePointSize(trajectory); i++) {
    setT(trajectory, 1.0, i);
    setX(trajectory, 1.0, i);
    setY(trajectory, 2.0, i);
    setV(trajectory, 1.0, i);
    setTheta(trajectory, M_PI, i);
    setA(trajectory, 1.0, i);
    setBeta(trajectory, delta_front, delta_rear, distance_front_axle, distance_rear_axle, i);
    setDeltaFront(trajectory, M_PI, i);
    setDeltaRear(trajectory, M_PI, i);
    setS(trajectory, 1.0, i);
  }

  gm::TransformStamped tf;
  tf.transform.translation.x = 10.0;
  tf.transform.translation.y = 20.0;
  tf.transform.translation.z = 30.0;
  tf.transform.rotation.x = 0.0;
  tf.transform.rotation.y = 0.0;
  tf.transform.rotation.z = 1.0;
  tf.transform.rotation.w = 0.0;

  tf2::doTransform(trajectory, trajectory_tf, tf);

  for (int i = 0; i < getSamplePointSize(trajectory); i++) {

    // transformed state
    EXPECT_NEAR(getX(trajectory_tf, i), 9.0, EPS);
    EXPECT_NEAR(getY(trajectory_tf, i), 18.0, EPS);
    EXPECT_NEAR(getTheta(trajectory_tf, i), 0.0, EPS);

    // transform-invariant state
    EXPECT_NEAR(getV(trajectory_tf, i), 1.0, EPS);
    EXPECT_NEAR(getA(trajectory_tf, i), 1.0, EPS);
    EXPECT_NEAR(getBeta(trajectory_tf, i), beta, EPS);
    EXPECT_NEAR(getDeltaFront(trajectory_tf, i), M_PI, EPS);
    EXPECT_NEAR(getDeltaRear(trajectory_tf, i), M_PI, EPS);
    EXPECT_NEAR(getS(trajectory_tf, i), 1.0, EPS);
  }
}


int main(int argc, char *argv[]) {
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: tf2_trajectory_planning_msgs/test/test_tf2_trajectory_planning_msgs.py
# ============================================================================
# MIT License
# 
# Copyright (c) 2025 Institute for Automotive Engineering (ika), RWTH Aachen University
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# ============================================================================

import math

import pytest
import tf2_geometry_msgs
import tf_transformations
from geometry_msgs.msg import PoseStamped, TransformStamped
from trajectory_planning_msgs.msg import DRIVABLE, Trajectory

from tf2_trajectory_planning_msgs import do_transform_trajectory
from trajectory_planning_msgs_utils.init import initialize_trajectory
from trajectory_planning_msgs_utils.state_getters import (
    get_theta_from_trajectory,
    get_t_from_trajectory,
    get_x_from_trajectory,
    get_y_from_trajectory,
)
from trajectory_planning_msgs_utils.state_setters import (
    set_theta_in_trajectory,
    set_t_in_trajectory,
    set_x_in_trajectory,
    set_y_in_trajectory,
)

EPS = 1e-9


def _stamp_to_sec(stamp):
    return float(stamp.sec) + float(stamp.nanosec) * 1e-9


def test_do_transform_trajectory_drivable():
    trajectory = Trajectory()
    trajectory.header.frame_id = "source"
    trajectory.header.stamp.sec = 1
    trajectory.header.stamp.nanosec = 200_000_000
    trajectory.type_id = DRIVABLE.TYPE_ID

    initialize_trajectory(trajectory, DRIVABLE.TYPE_ID, 1)

    set_t_in_trajectory(trajectory, 0.5, 0)
    set_x_in_trajectory(trajectory, 1.0, 0)
    set_y_in_trajectory(trajectory, -2.0, 0)
    set_theta_in_trajectory(trajectory, 0.25, 0)

    transform = TransformStamped()
    transform.header.frame_id = "target"
    transform.child_frame_id = "source"
    transform.header.stamp.sec = 3
    transform.header.stamp.nanosec = 700_000_000
    transform.transform.translation.x = 4.0
    transform.transform.translation.y = -1.0
    transform.transform.translation.z = 0.0
    q = tf_transformations.quaternion_from_euler(0.0, 0.0, math.pi / 3.0)
    transform.transform.rotation.x = q[0]
    transform.transform.rotation.y = q[1]
    transform.transform.rotation.z = q[2]
    transform.transform.rotation.w = q[3]

    transformed = do_transform_trajectory(trajectory, transform)

    assert transformed.header.frame_id == transform.header.frame_id
    assert transformed.header.stamp == transform.header.stamp

    pose_in = PoseStamped()
    pose_in.header.frame_id = trajectory.header.frame_id
    pose_in.pose.position.x = 1.0
    pose_in.pose.position.y = -2.0
    pose_in.pose.position.z = 0.0
    q_in = tf_transformations.quaternion_from_euler(0.0, 0.0, 0.25)
    pose_in.pose.orientation.x = q_in[0]
    pose_in.pose.orientation.y = q_in[1]
    pose_in.pose.orientation.z = q_in[2]
    pose_in.pose.orientation.w = q_in[3]
    pose_expected = tf2_geometry_msgs.do_transform_pose(pose_in, transform)

    assert math.isclose(
        get_x_from_trajectory(transformed, 0),
        pose_expected.pose.position.x,
        rel_tol=EPS,
        abs_tol=EPS,
    )
    assert math.isclose(
        get_y_from_trajectory(transformed, 0),
        pose_expected.pose.position.y,
        rel_tol=EPS,
        abs_tol=EPS,
    )

    expected_yaw = tf_transformations.euler_from_quaternion(
        [
            pose_expected.pose.orientation.x,
            pose_expected.pose.orientation.y,
            pose_expected.pose.orientation.z,
            pose_expected.pose.orientation.w,
        ]
    )[2]
    assert math.isclose(
        get_theta_from_trajectory(transformed, 0),
        expected_yaw,
        rel_tol=EPS,
        abs_tol=EPS,
    )

    time_offset = _stamp_to_sec(transform.header.stamp) - _stamp_to_sec(trajectory.header.stamp)
    assert math.isclose(
        get_t_from_trajectory(transformed, 0),
        0.5 + time_offset,
        rel_tol=EPS,
        abs_tol=EPS,
    )


if __name__ == "__main__":
    pytest.main()
